sim/firesim-lib/src/main/cc/bridges/dromajo.cc

```diff
@@ -0,0 +1,305 @@
+        "--clint",
+        DROMAJO_CLINT_BASE ":" DROMAJO_CLINT_SIZE,
+        "--memory_size",
+        DROMAJO_MEM_SIZE,
+        "--save",
+        "dromajo_snap",
+        "--dtb",
+        (char*)this->dromajo_dtb.c_str(),
+        (char*)this->dromajo_bin.c_str()
+    };
+
+    if (MAX_ARGS < local_argc) {
+        printf("[DRJ_ERR] Too many arguments\n");
+        exit(1);
+    }
+
+    printf("[INFO] Dromajo command: \n");
+    for(int i = 0; i < local_argc; ++i) {
+        printf("%s ", local_argv[i]);
+    }
+    printf("\n");
+
+    this->dromajo_state = dromajo_cosim_init(local_argc, local_argv);
+    if (this->dromajo_state == NULL) {
+        printf("[ERROR] Error setting up Dromajo\n");
+        exit(1);
+    }
+}
+
+/**
+ * Call Dromajo co-sim functions with an aligned buffer.
+ * This returns the return code of the co-sim functions.
+ */
+int dromajo_t::invoke_dromajo(uint8_t* buf) {
+    bool valid = buf[0];
+    int hartid = 0; // only works for single core
+    // this crazy to extract the right value then sign extend within the size
+    uint64_t pc = ((int64_t)(*((int64_t*)(buf + this->_iaddr_offset)) & BIT_MASK(uint64_t, this->_iaddr_width*8)) << (sizeof(uint64_t) - this->_iaddr_width)*8) >> (sizeof(uint64_t) - this->_iaddr_width)*8;
+    uint32_t insn = ((int32_t)(*((int32_t*)(buf + this->_insn_offset)) & BIT_MASK(uint32_t, this->_insn_width*8)) << (sizeof(uint32_t) - this->_insn_width)*8) >> (sizeof(uint32_t) - this->_insn_width)*8;
+    uint64_t wdata = ((int64_t)(*((int64_t*)(buf + this->_wdata_offset)) & BIT_MASK(uint64_t, this->_wdata_width*8)) << (sizeof(uint64_t) - this->_wdata_width)*8) >> (sizeof(uint64_t) - this->_wdata_width)*8;
+    uint64_t mstatus = 0; // default not checked
+    bool check = true; // default check all
+    bool interrupt = buf[this->_interrupt_offset];
+    bool exception = buf[this->_exception_offset];
+    int64_t cause = (*((uint64_t*)(buf + this->_cause_offset)) & BIT_MASK(uint64_t, this->_cause_width*8)) | ((uint64_t)interrupt << 63);
+
+    if (valid) {
+#ifdef DEBUG
+        if (interrupt || exception)
+            fprintf(stderr, "[DEBUG] INT/EXCEP raised (on valid): cause = %lx\n", cause);
+#endif
+        return dromajo_cosim_step(this->dromajo_state, hartid, pc, insn, wdata, mstatus, check);
+    }
+
+    if ((interrupt || exception) && !(this->saw_int_excp)) {
+#ifdef DEBUG
+        fprintf(stderr, "[DEBUG] INT/EXCEP raised (normal): cause = %lx\n", cause);
+#endif
+        dromajo_cosim_raise_trap(this->dromajo_state, hartid, cause);
+        this->saw_int_excp = true;
+    }
+
+    return 0;
+}
+
+/**
+ * Read queue and co-simulate
+ */
+void dromajo_t::process_tokens(int num_beats) {
+    // TODO: as opt can mmap file and just load directly into it.
+    alignas(4096) uint8_t OUTBUF[QUEUE_DEPTH * sizeof(uint64_t) * 8];
+    pull(this->_dma_addr, (char*)OUTBUF, num_beats * sizeof(uint64_t) * 8);
+
+    // skip if co-sim not enabled
+    if (!this->dromajo_cosim) return;
```

Comment:
Note: TracerV has a similar disable. But before the queue is drained... thus if TracerV is disabled... wouldn't the queue fill up and stall the simulation? Maybe I'm missing something here.

Reply:
Sorry for the late reply, but TracerV puts the bridge in a state where it will drop all tokens to prevent you correct observation from happening.